kroxylicious/src/main/templates/Kproxy/PerMessageInvoker.ftl

```diff
@@ -0,0 +1,116 @@
+<#--
+
+    Copyright Kroxylicious Authors.
+
+    Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+
+-->
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ${outputPackage};
+
+import io.kroxylicious.proxy.frame.DecodedRequestFrame;
+import io.kroxylicious.proxy.frame.DecodedResponseFrame;
+<#list messageSpecs as messageSpec>
+import org.apache.kafka.common.message.${messageSpec.name}Data;
+</#list>
+import org.apache.kafka.common.protocol.ApiKeys;
+
+import java.util.HashSet;
+import java.util.EnumSet;
+import java.util.Set;
+
+public class PerMessageInvoker implements KrpcFilter
```

Comment:
I don't think the invocation of a filter needs to be conflated with the invoker being a filter itself. Or at least, the benefits of using inheritance aren't clear to me. Using inheritance like this, while convenient for the current code base, means we can't make KrpcFilter a `sealed` interface in the future without also exposing `PerMessageInvoker` in the `given` clause, but I definitely see `PerMessageInvoker` as an implementation detail.

Reply:
Okay, so I'm hearing we should reserve the KrpcFilter hierarchy for the filters. Maybe the Invoker should have a separate interface though so that we can build alternative invokers (like for the RequestFilter, ResponseFilter idea discussed).